fix(core): standard list a11y

dmitry-stepanenko opened the pull request:
## Related Issue(s)

<!-- If this PR fixes multiple issues, please use the full syntax(`closes #issue-number`) for each issue so that each issue gets automatically closed on PR merge; for example: `closes #0001, closes #0002`, and so on. -->

closes https://github.com/SAP/fundamental-ngx/issues/6222

## Description
Overview of changes that were made:
Issue 1 - done
Issue 2 - done
Issue 3 - done
Issue 4 - **not done**, now group header is not focused at all, header information is conveyed when item is focused
Issue 5 - done
Issue 6 - done
Issue 7 - done
Issue 8 - done
Issue 9 - done
Issue 10 - done
Issue 11 - done
Issue 12 - **not done**, such functionality requires significant changes and new behavior should be described by specs. Currently there’s not nothing about keyboard support there
<!-- Enter short description of the change -->

## Screenshots

<!-- If you've made any style changes, please provide appropriate screenshots (before and after) to help reviewers. -->

### Before:

### After:

#### Please check whether the PR fulfills the following requirements

##### During Implementation

1. Visual Testing:

-   [x] visual misalignments/updates
-   [x] check Light/Dark/HCB/HCW themes
-   [x] RTL/LTR - proper rendering and labeling
-   [n/a] responsiveness(resize)
-   [n/a] Content Density (Cozy/Compact/(Condensed))
-   [x] States - hover/disabled/focused/active/on click/selected/selected hover/press state
-   [x] Interaction/Animation - open/close, expand/collapse, add/remove, check/uncheck
-   [x] Mouse vs. Keyboard support
-   [n/a] Text Truncation

2. API and functional correctness

-   [x] check for console logs (warnings, errors)
-   [x] API boundary values
-   [n/a] different combinations of components - free style
-   [n/a] change the API values during testing

3. Documentation and Example validations

-   [n/a] missing API documentation or it is not understandable
-   [n/a] poor examples
-   [n/a] Stackblitz works for all examples

4. Accessibility testing
5. Browser Testing - Edge, Safari, Chrome, Firefox

##### PR Quality

-   [x] the commit message(s) follows the guideline:
        https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
-   [n/a] tests for the changes that have been done
-   [x] all items on the PR Review Checklist are addressed :
        https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
-   [n/a] Run npm run build-pack-library and test in external application
-   [n/a] update `README.md`
-   [n/a] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)

InnaAtanasova commented on apps/docs/src/app/core/component-docs/list/examples/list-action-example/list-action-example.component.html:
```diff
@@ -5,7 +5,7 @@
         <span fd-list-title> List Item {{ i }} </span>
     </li>
-    <li fd-list-item [action]="true" [growing]="true">
+    <li fd-list-item [action]="true" [growing]="true" tabindex="-1" aria-hidden="true">
```
Please explain in more details why you did this, esp the part where you add tabindex to the `li` element. The content of the li element is a button that takes the whole width and height and is by default focussable and what concerns the user is the content inside the button, so I really struggle to understand why we need to focus the `li` element.

dmitry-stepanenko replied:
This list-item element has a button inside, both list-item and inner button can receive focus. Because of this you have to actually hit TAB 2 times to go through this element. So I've added these attributes in order to omit the focus on list-item. 

The problem is currently it's supported to add multiple elements inside the tab, while it's possible that none of the descendants will be focusable. So not sure what would be the better solution

mikerodonnell89 replied:
> 

In which examples does the list-item element have a button inside? Not in this one, and only one tab press is needed to move through the items on the current implementation https://fundamental-ngx.netlify.app/#/core/list#action

dmitry-stepanenko replied:
@mikerodonnell89 you can see it in this `Action list` example. 


https://user-images.githubusercontent.com/33101123/145172327-d86c297b-463e-49ec-b20a-96c6749b1c5f.mp4



However, I removed `aria-hidden` now, since this change is so questionable. It was added in order to not include list item with button in the total number of items.

platon-rov replied:
There is not much sence to have li element focusable at all, maybe we should remove `tabindex=0` from there?

dmitry-stepanenko replied:
@platon-rov don't think so: we're not limiting the content of `fd-list-item` to be something specific, so it can be something that is not focusable. Because of this the behavior of the list might be not that predictable

platon-rov replied:
@dmitry-stepanenko okay, so let' imagine we have only text in `fd-list-item` then for what reason we need `tabindex=0` on that item?

platon-rov replied:
I saw the pr that makes only first item focusable and next are accessible using arrows but cannot find it right now.

platon-rov replied:
Ahaha, that's current pr, that's the reason i didn't find it :D

InnaAtanasova commented on apps/docs/src/app/core/component-docs/list/examples/list-data-example/list-data-example.component.html:
```diff
@@ -12,16 +13,28 @@
         </span>
         <span fd-list-secondary>
-            <button fd-button compact="true" glyph="delete" (click)="removeItem(ind)"></button>
+            <button ariaLabel="Remove" fd-button compact="true" glyph="delete" (click)="removeItem(ind)"></button>
```
this is a native html element, a button. Why don't you use `aria-label`? Why did you decide go for an input property?

dmitry-stepanenko replied:
There's a host binding in `fd-button` for aria-label, that can override plain `aria-label` attribute

dmitry-stepanenko commented on libs/core/src/lib/list/list-item/list-item.component.ts:
```diff
@@ -11,24 +11,28 @@ import {
 import { ListFocusItem } from '../list-focus-item.model';
 import { ButtonComponent } from '@fundamental-ngx/core/button';
 
+let UNIQUE_ID = 0;
```
done

dimamarksman commented:
Focusable <li/> by default is really confusing cause it forces to hit Tad twice if there is some focusable element inside this list item.
I believe interactive controls can't have interactive children.
If user wants to make <li> interactive and assign some event handler to it then the user needs to add tabindex="0"

dmitry-stepanenko commented:
@dimamarksman @platon-rov regarding the focusable list items: I agree with the fact it's confusing to have it and instead there should only be focusable elements inside the list, if any. 

However, I can see the behavior with focusable list items on SAPUI5 (using arrow keys) https://sapui5.hana.ondemand.com/#/entity/sap.m.List/sample/sap.m.sample.ListCounter

Maybe the right approach would be to keep tab for actually traversing focusable elements and have arrow keys to go through the list as we can see on SAPUI5?

platon-rov commented on libs/core/src/lib/list/list-item/list-item.component.scss:
```diff
@@ -0,0 +1,12 @@
+    width: 1px;
+    overflow: hidden;
+    /* IE 7+ only support clip without commas */
+    clip: rect(1px 1px 1px 1px);
```
We don't care about IE anymore.

platon-rov commented on libs/core/src/lib/list/list-item/list-item.component.ts:
```diff
@@ -124,14 +154,35 @@ export class ListItemComponent extends ListFocusItem implements AfterContentInit
+        return describedBy;
+    }
+
+    readonly uniqueId = 'fdp-list-item-' + ++listItemUniqueId;
```
`@hidden`?

platon-rov reviewed (commented):
Okay, that's makes more sense to have items not focusable but keyboard accessible with arrow buttons.

There is strange behavior in action list example, where "show more" button is shown, after pressing tab not first but show more element gets focus, then pressing arrows behaves strange.

https://user-images.githubusercontent.com/20265336/149318105-8a8885f8-bdaf-4aa2-b4ea-3b789cc563b7.mov

dmitry-stepanenko commented:
> Okay, that's makes more sense to have items not focusable but keyboard accessible with arrow buttons.
> 
> There is strange behavior in action list example, where "show more" button is shown, after pressing tab not first but show more element gets focus, then pressing arrows behaves strange.
> 
>  Screen.Recording.2022-01-13.at.12.59.13.mov

@platon-rov thanks for catching this, looks like this issue was always there. It was related to tracking only direct focus on the list element. And when you tabbing over the element, focus was applied to the button, thus no event received. Because of this keymanager had invalid active item index.

dmitry-stepanenko commented:
@mikerodonnell89 @InnaAtanasova please review again

Veerakarthick commented:
issue1:Partally fixed
Observed Behavior:
When accessing the list items state selected is read out but the user is not able to select the list item using 'Space bar (or) Enter key.

Expected Behavior:
Ideally, the user should be able to select the list item using the keyboard keys.
screenshot:
![image](https://user-images.githubusercontent.com/37998745/152917494-fb890070-1dd4-4de1-9a20-29e1407d6abf.png)

Issue2: Not Reproducible

Issue3:
Observed Behavior:
Item count is read for group header also within list.

Expected Behavior:
Ideally, item count should not be read for group headers.
Screenshot:
![image](https://user-images.githubusercontent.com/37998745/152920238-cc66c2c2-37ba-4013-b9e9-395bbb26e630.png)

Issue4: Fixed.

Issue5: Fixed.

Issue6:
Observed Behavior:
Text alternative defining the purpose of UI is not provided for buttons highlighted in red.

Speech output is:
Button
To activate press spacebar.
Button
To activate press spacebar.

Expected Behavior:
Ideally, text alternative describing the purpose of UI should be provided for each UI.
Screenshot:
![image](https://user-images.githubusercontent.com/37998745/152920608-cad2839b-8e67-427d-88ad-d3b5585249b9.png)

Issue7:
Observed Behavior:
After pressing the buttons, information such as ascending/descending is not read.

Speech output is:
Button
To activate press spacebar.
Button
To activate press spacebar.

Expected Behavior:
Ideally, information such as ascending/descending should be conveyed.

Screenshot:
![image](https://user-images.githubusercontent.com/37998745/152921344-e0a2c6d8-b85b-4918-8381-b1305ad74333.png)


Issue8: Fixed.

Issue9:
Observed Behavior:
Text alternative defining the purpose of UI is not provided for Buttons within the list.

Speech output is:
list
with 5 items
Apple
1 of 5
To move to an item press the Arrow keys.
Button
To activate press spacebar.

Expected Behavior:
Ideally, text alternative describing the purpose of UI should be provided for each UI.

Screenshot:
![image](https://user-images.githubusercontent.com/37998745/152921618-50aba22b-2ffd-45d0-ac00-368c824b304d.png)

issue10:
Observed Behavior:
Group headers are not read out when list items are accessed.

Speech output is:
list
with 11 items
List item 1
2 of 11
To move to an item press the Arrow keys.
List item 2
3 of 11
To move to an item press the Arrow keys.
List item 3
4 of 11
To move to an item press the Arrow keys.
List item 4
5 of 11
To move to an item press the Arrow keys.
List item 5
7 of 11
To move to an item press the Arrow keys.
List item 6
8 of 11
To move to an item press the Arrow keys.
List item 7
9 of 11
To move to an item press the Arrow keys.
List item 8
10 of 11
To move to an item press the Arrow keys.

Expected Behavior:
Ideally, corresponding group headers should be read out.

Screenshot:
![image](https://user-images.githubusercontent.com/37998745/152922148-a8adbd02-d9cf-4f3d-9387-204aeb18762e.png)

Issue11: Fixed.

Issue12:
Observed Behavior:
Not possible to drag and drop list items via keyboard.

Expected Behavior:
Ideally, it should be possible to drag and drop list items using keyboard.
screenshot:
![image](https://user-images.githubusercontent.com/37998745/152922514-de48edc2-99f6-4e9c-a0e3-6ee8506d6ffb.png)


Retest Environment:
URL: https://deploy-preview-7308--fundamental-ngx.netlify.app/
Browser: Chrome Version 96.0.4664.110 (Official Build) (64-bit)
Screen Reader: JAWS 2021.2111.13 ILM

Regards,
Karthikeyan.c

dmitry-stepanenko commented:
@Veerakarthick responding to your comment above:
issue1 - made changes to support keyboard selection for interactive lists https://deploy-preview-7308--fundamental-ngx.netlify.app/#/core/list
issue 3 - this is a side effect of subheader being part of the list and not really avoidable, if we want them to be focusable
issue 6 - buttons have aria label now and are read properly. Verified in JAWS and VoiceOver
issue 7 - buttons have aria label now and are read properly. The issue with them not being read after clicking looks like an issue with JAWS not announcing the click target
issue 9 - buttons have aria label now and are read properly. Verified in JAWS and VoiceOver
issue 10 - please retest, it's working for me. Verified in JAWS and VoiceOver

Veerakarthick commented:
Hello Colleague,

With the below preview URL am not able to find out the component 'Standard list' please check and guide me to find out the component?

URL: https://fundamental-ngx-gh--pr7308-ds-6222-standard-lis-0d8ycow7.web.app/

screenshot:
![image](https://user-images.githubusercontent.com/37998745/155260981-7e23b857-6c4f-4dc4-983e-f945fc3bebe8.png)


Thanks and Regards,
Karthikeyan.c

dmitry-stepanenko commented:
@Veerakarthick please switch to platform components 
<img width="503" alt="image" src="https://user-images.githubusercontent.com/33101123/155277436-d7ed9263-4ee1-4d88-9666-2ff698ba4313.png">

Veerakarthick commented:
Hello Colleague,

The component is belong to 'Core' component and the results are below:

issue1: Fixed
Issue2: Not Reproducible
Issue3:
Observed Behavior:
Item count is read for group header also within list.
Expected Behavior:
Ideally, item count should not be read for group headers.
Screenshot:
 
Issue4: Fixed.
Issue5: Fixed.
Issue6:
Observed Behavior:
Text alternative defining the purpose of UI is not provided for buttons highlighted in red.
Speech output is:
Button
To activate press spacebar.
Button
To activate press spacebar.
Expected Behavior:
Ideally, text alternative describing the purpose of UI should be provided for each UI.
Screenshot:
 
Issue7:
Observed Behavior:
After pressing the buttons, information such as ascending/descending is not read.
Speech output is:
Button
To activate press spacebar.
Button
To activate press spacebar.
Expected Behavior:
Ideally, information such as ascending/descending should be conveyed.
Screenshot:
 
Issue8: Fixed.
Issue9:
Observed Behavior:
Text alternative defining the purpose of UI is not provided for Buttons within the list.
Speech output is:
list
with 5 items
Apple
1 of 5
To move to an item press the Arrow keys.
Button
To activate press spacebar.
Expected Behavior:
Ideally, text alternative describing the purpose of UI should be provided for each UI.
Screenshot:
 
issue10: Fixed

Issue11: Fixed.
Issue12:
Observed Behavior:
Not possible to drag and drop list items via keyboard.
Expected Behavior:
Ideally, it should be possible to drag and drop list items using keyboard.
screenshot:
 
Retest Environment:
URL: https://fundamental-ngx-gh--pr7308-ds-6222-standard-lis-n5r0ii9f.web.app/
Browser: Chrome Version 96.0.4664.110 (Official Build) (64-bit)
Screen Reader: JAWS 2021.2111.13 ILM

Regards,
Karthikeyan.c

Veerakarthick commented:
Hello Colleague,
Retest has been taken status are below:
issue1: Fixed
Issue2: Not Reproducible
Issue3:
Observed Behaviour:
Item count is read for group header also within list.
 
Expected Behaviour:
Ideally, item count should not be read for group headers.
Screenshot:
Issue4: Fixed.
Issue5: Fixed.
Issue6: partially fixed
Observed Behaviour:
Text alternative defining the purpose of UI is not provided for buttons highlighted in red.
Speech output is:
Sort ascending  Button  
To activate press spacebar.
Sort descending  Button  
To activate press spacebar..
 
Expected Behavior:
Ideally, text alternative describing the purpose of UI should be provided for each UI.
Screenshot:
Issue7:
Observed Behavior:
After pressing the buttons, information such as ascending/descending is not read.
Speech output is:
Sort ascending  Button  
To activate press spacebar.
Sort descending  Button  
To activate press spacebar.

 
Expected Behavior:
Ideally, information such as ascending/descending should be conveyed.
Screenshot:
Issue8: Fixed.
Issue9: Partially fixed
Observed Behavior:
Text alternative defining the purpose of UI is not provided for Buttons within the list.
Speech output is:.
Remove  Button  
To activate press spacebar.
Remove  Button  
To activate press spacebar.
 
Expected Behavior:
Ideally, text alternative describing the purpose of UI should be provided for each UI.
Screenshot:
issue10: Fixed
Issue11: 
Observed Behavior:
Not possible to drag and drop list items via keyboard.
Expected Behavior:
Ideally, it should be possible to drag and drop list items using keyboard.
screenshot:
Retest Environment:
Browse the preview: https://deploy-preview-7308--fundamental-ngx.netlify.app/
Browser: Chrome Version Version 100.0.4896.75 (Official Build) (64-bit)
Screen Reader: JAWS 2021.2111.13 ILM
Regards,
Karthikeyan.c

dmitry-stepanenko commented:
@Veerakarthick I think I understand now: you were talking about the text titles of the elements, not how they are read in issues 6 and 9. Fixed those. Also added live announcing according to issue 7.

According to your last review there are few things left:
issue 3 - this is a side effect of subheader being part of the list and not really avoidable, if we want them to be focusable
issue 12 - such functionality requires significant changes and new behavior should be described by specs. Currently there’s not nothing about keyboard support there

I suggest to skip issues 3 and 12.

Karthickc1994 commented:
Hello Colleague,

Retest has been taken within the below environment  test results are below:

issue1: Fixed,
issue2: Fixed.
Issue3: Still reproducible (Need input from PO)
Group headers are added in the list item count.
Reading made as :  list  
with 8 items
List Group Header    
1 of 8
To move to an item press the Arrow keys.
Screen shot:
![image](https://user-images.githubusercontent.com/105037808/167366559-c0ae5be9-f683-4082-b7e3-866819cf6230.png)


issue4: Fixed.
Issue5: Fixed.
Issue6: Fixed.
Issue7: Fixed.
Issue8: Fixed.
Issue9: Fixed.
Issue10: Fixed.
Issue11: Fixed.
Issue 12: Still reproducible (Need input from PO)
Observed Behavior:
Not possible to drag and drop list items via keyboard.


Retest Environment:

URL: https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist 
Browser: Version 101.0.4951.54 (Official Build) (64-bit)
Screen reader: JAWS 2022.2204. 20ILM

Thanks and Regards,
Karthikeyan.c

droshev commented:
@dmitry-stepanenko thanks for the work! go ahead.
